OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,364 @@
+
+			Initialize();
+
+			Game.OnQuit += Disconnect;
+
+			try
+			{
+				AddChatLine("Connecting to {0}...".F(Game.Settings.Irc.Hostname));
+				Irc.Client.Connect(Game.Settings.Irc.Hostname, Game.Settings.Irc.Port);
+			}
+			catch (Exception e)
+			{
+				AddChatLine("Connection error: {0}".F(e.Message));
+				Log.Write("irc", e.ToString());
+			}
+
+			Irc.ListenThread = new Thread(new ThreadStart(Listen));
+			Irc.ListenThread.Name = "IrcListenThread";
+			Irc.ListenThread.Start();
+		}
+
+		void OnPong(object sender, PongEventArgs e)
+		{
+			if (pingSent)
+			{
+				AddChatLine("PONG recieved after {0} ms.".F(e.Lag.Milliseconds));
+				pingSent = false;
+			}
+			else
+				Log.Write("irc", "PONG sent after {0} ms.".F(e.Lag.Milliseconds));
+		}
+
+		void Listen()
+		{
+			while (Irc.Client != null)
+				Irc.Client.ListenOnce();
+		}
+
+		Widget MakeLabelWidget(LabelWidget template, string item)
+		{
+			var widget = (LabelWidget)template.Clone();
+			var font = Game.Renderer.Fonts[widget.Font];
+			item = WidgetUtils.WrapText(item, widget.Bounds.Width, font);
+			widget.Bounds.Height = font.Measure(item).Y;
+			widget.GetText = () => item;
+			widget.Id = item;
+			return widget;
+		}
+
+		void AddChatLine(string text)
+		{
+			Log.Write("irc", text);
+			Game.RunAfterTick(() =>
+			{
+				var scrolledToBottom = historyPanel.ScrolledToBottom;
+
+				var newChild = MakeLabelWidget(historyTemplate, text);
+				historyPanel.AddChild(newChild);
+
+				if (scrolledToBottom)
+					historyPanel.ScrollToBottom(smooth: true);
+			});
+		}
+
+		bool EnterPressed()
+		{
+			if (!inputBox.Text.Any())
+				return true;
+
+			var text = inputBox.Text;
+			inputBox.Text = "";
+
+			if (text.StartsWith("/nick "))
+			{
+				var nick = text.Replace("/nick ", string.Empty);
+				if (Rfc2812.IsValidNickname(nick))
+					Irc.Client.RfcNick(nick);
+				else
+					AddChatLine("Invalid nickname.");
+			}
+			else if (text.StartsWith("/ping "))
+			{
+				Irc.Client.RfcPing(Irc.Client.GetIrcUser(text.Replace("/ping ", string.Empty)).Host);
+				pingSent = true;
+			}
+			else if (text.StartsWith("/"))
+				AddChatLine("Unknown command.");
+			else
+			{
+				AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), Irc.Client.Nickname, text));
+				Irc.Client.SendMessage(SendType.Message, "#" + Game.Settings.Irc.Channel, text);
+			}
+
+			return true;
+		}
+
+		void OnConnected(object sender, EventArgs e)
+		{
+			AddChatLine("Connected.");
+
+			if (!Rfc2812.IsValidNickname(Game.Settings.Irc.Nickname))
+			{
+				AddChatLine("Invalid nickname. Can't login.");
+				return;
+			}
+
+			Irc.Client.Login(new[] { Game.Settings.Irc.Nickname }, "in-game IRC client", 0, "OpenRA");
+
+			Irc.Client.RfcJoin("#" + Game.Settings.Irc.Channel);
+		}
+
+		void OnError(object sender, ErrorEventArgs e)
+		{
+			AddChatLine("Error: ".F(e.ErrorMessage));
+			Log.Write("irc", "Error: {0}".F(e.ErrorMessage));
```

Comment:
Our Log.Write isn't thread-safe - this is an issue with Log.Write rather than the IRC stuff but we'd still need to rectify it if we're going to start calling it from the IRC thread.